pipeline/psth_foraging.py

```diff
@@ -0,0 +1,602 @@
+
+        stim_attrs = set((experiment.Photostim * experiment.PhotostimBrainRegion
+                          * experiment.PhotostimEvent).heading.names) - set(experiment.Session.heading.names)
+        behav_attrs = set((experiment.BehaviorTrial * experiment.WaterPortChoice).heading.names)
+
+        _stim_key = {k: v for k, v in _restr.items() if k in stim_attrs}
+        _behav_key = {k: v for k, v in _restr.items() if k in behav_attrs}
+
+        stim_key = {k: v for k, v in restr.items() if k in stim_attrs}
+        behav_key = {k: v for k, v in restr.items() if k in behav_attrs}
+
+        return (((experiment.BehaviorTrial * experiment.WaterPortChoice & behav_key) - [{k: v} for k, v in _behav_key.items()]) -
+                ((experiment.PhotostimEvent * experiment.PhotostimBrainRegion * experiment.Photostim & stim_key)
+                 - [{k: v} for k, v in _stim_key.items()]).proj())
+
+    @classmethod
+    def _get_trials_include_stim(cls, **kwargs):
+        # Note: inclusion (attr) is AND - exclusion (_attr) is OR
+        log.debug('_get_trials_include_stim: {}'.format(kwargs))
+
+        restr, _restr = {}, {}
+        for k, v in kwargs.items():
+            if k.startswith('_'):
+                _restr[k[1:]] = v
+            else:
+                restr[k] = v
+
+        stim_attrs = set((experiment.Photostim * experiment.PhotostimBrainRegion
+                          * experiment.PhotostimEvent).heading.names) - set(experiment.Session.heading.names)
+        behav_attrs = set((experiment.BehaviorTrial * experiment.WaterPortChoice).heading.names)
+
+        _stim_key = {k: v for k, v in _restr.items() if k in stim_attrs}
+        _behav_key = {k: v for k, v in _restr.items() if k in behav_attrs}
+
+        stim_key = {k: v for k, v in restr.items() if k in stim_attrs}
+        behav_key = {k: v for k, v in restr.items() if k in behav_attrs}
+
+        return (((experiment.BehaviorTrial * experiment.WaterPortChoice & behav_key) - [{k: v} for k, v in _behav_key.items()]) &
+                ((experiment.PhotostimEvent * experiment.PhotostimBrainRegion * experiment.Photostim & stim_key)
+                 - [{k: v} for k, v in _stim_key.items()]).proj())
+
+
+@schema
+class UnitPsth(dj.Computed):
+    definition = """
+    -> TrialCondition
+    -> ephys.Unit
+    ---
+    unit_psth=NULL: longblob
+    """
+    psth_params = {'xmin': -3, 'xmax': 3, 'binsize': 0.04}
+
+    @property
+    def key_source(self):
+        """
+        For those conditions that include stim, process those with PhotostimBrainRegion already computed only
+        Only units not of type "all"
+        """
+        nostim = (ephys.Unit * (TrialCondition & 'trial_condition_func = "_get_trials_exclude_stim"')
+                  & 'unit_quality != "all"')
+        stim = ((ephys.Unit & (experiment.Session & experiment.PhotostimBrainRegion))
+                * (TrialCondition & 'trial_condition_func = "_get_trials_include_stim"') & 'unit_quality != "all"')
```

Comment:
I just realized another thing: should I add a restriction to `key_source` here to make sure `psth_foraging.UnitPsth` only applies to the foraging task? and vice versa for `psth.UnitPsth`?

Reply:
yes, indeed. Perhaps something on the `return` line (line 290):

```
return nostim.proj() + stim.proj() & (experiment.Session & (experiment.BehaviorTrial & 'task LIKE "foraging%"'))
```